### DISCUSSION TITLE
Any good resources for how to create as custom Node?

### DISCUSSION THREAD

**Discussion by ChrisFlick** [comfyanonymous/ComfyUI#1291] [General]
scrounging the internet and finding very little to none

**Comment by coreyryanhanson** [comment]
@ChrisFlick As far as I know there isn't really anything comprehensive.

Luckily there is an [example file](https://github.com/comfyanonymous/ComfyUI/blob/master/custom_nodes/example_node.py.example) in the custom nodes folder. The comments within the example code are very helpful. If you have Python knowledge, between that and the many examples out there in the wild, you should be good to go. Once you get a basic node working, you can also use javascript to enhance it.

Things you need for a basic working node.
* A .py file with:
  * A node defined by a properly formatted class
  * NODE_CLASS_MAPPINGS: A dictionary that maps the name for the node (a string that should not overlap existing nodes) and the class
  * Optionally NODE_DISPLAY_NAME_MAPPINGS will allow a more human readable name to show up in the context menu if it is also provided.
* An __init__.py file. This tells ComfyUI where the previously mentioned mapping dictionaries are located. It can be very simple with very few lines. Look at other custom nodes to get the basic idea.

Optional:
* A requirements.txt file if there are any additional dependencies that your node will need.

That's basically it. 2 to 3 files in a folder will give you a good start. Then you can start experimenting with adding complexity.

**Reply by SmokeyRGB** [reply]
> Once you get a basic node working, you can also use javascript to enhance it.

its this I am having trouble with understanding. I am understanding the basic principle of making custom nodes but it would be really helpful to have any resources on how to integrate javascript into it...
I would love to make a simple node just changing the document title to a given String.

**Reply by el0911** [reply]
ok am gonna see if i fine tune a gpt model on this and tell it to make nodes for me

**Comment by blurymind** [comment]
Any simple examples of how to render custom html inside a node? Can one put a reactjs component inside comfyui node?

I have something in mind to improve on one of the existing nodes usability wise

**Comment by wibur0620** [comment]
https://docs.comfy.org/essentials/custom_node_walkthrough
I don't know if this will be helpful to you.

**Comment by Nn0wk** [comment]
https://github.com/Suzie1/ComfyUI_Guide_To_Making_Custom_Nodes

**Reply by Nn0wk** [reply]
https://docs.comfy.org/custom-nodes/custom_node_server_overview

**Comment by ltdrdata** [comment]
https://docs.comfy.org/essentials/custom_node_overview

**Comment by schoenid** [comment]
Conclusion:
The provided reaources do not really help.

As example: 
There are keywords in a nodes class like DESCRIPTION, which are not documented.
Is it possible to define own keywords,like FAMILY or PACKAGE ?
Which other keywords do exist?

There is also a JavaScript part for nodes, which contains as example the setting `nodeData`.
How to get its definition or how to extend this?
For custom nodes this is mostly used for initialization of own nodes.
Some nodes are using it for the node names together with category.

Btw: is there a documentaion for categories?

**Comment by arcum42** [comment]
I'll throw this link in the mix:
https://github.com/comfyanonymous/ComfyUI/tree/master/comfy/comfy_types

If you look over at this file:
https://github.com/comfyanonymous/ComfyUI/blob/master/comfy/comfy_types/node_typing.py

You'll note information about DESCRIPTION, and other keywords.

You can also learn a fair amount by just looking at how base nodes are implemented.
https://github.com/comfyanonymous/ComfyUI/blob/master/nodes.py

CATEGORIES is simple enough. The node gets put in whatever category you name, and if you put in a '/', it'll subcategorize. 

For example, in the custom nodes I've been working on:
https://github.com/arcum42/ComfyUI_SageUtils

I put practically everything in "Sage Utils" first, so that all my nodes are in one place, then do subcategories, like ``` CATEGORY = "Sage Utils/text"```

Looking at other people's custom nodes and seeing how they implemented things can be pretty useful.

You might also find some of the example custom nodes over here helpful:
https://github.com/cozy-comfyui

**Comment by schoenid** [comment]
> [arcum42](https://github.com/arcum42)
> [5 hours ago](https://github.com/comfyanonymous/ComfyUI/discussions/1291#discussioncomment-12735435)
> I'll throw this link in the mix:
> https://github.com/comfyanonymous/ComfyUI/tree/master/comfy/comfy_types
> 
> If you look over at this file:
> https://github.com/comfyanonymous/ComfyUI/blob/master/comfy/comfy_types/node_typing.py
> 
> You'll note information about DESCRIPTION, and other keywords.
> 
> ...

Wow, great and thanks a lot !!!

Finally the main thing I was looking for is in the JavaScript class nodeData itself.
I was looking for an alternative to not to be required to use an own category for initialization.
Now I found, that `nodeData.python_module` contains all I need: `"custom_nodes.my_custom"`.
(found while debugging with browser DevTools ...)

I'm finally at the point, where I can do the initialization as follows:
```
app.registerExtension({
	name: "my_custom.output_widget",
	async beforeRegisterNodeDef(nodeType, nodeData, app) {
		if (nodeData.python_module != "custom_nodes.my_custom") {
			return;
		}
                ...
```

This way I'm not required to use the category and I'm free to define my nodes for "standard" categories, like "sampler", "viewer" or others.
I only still have to figure out, how these standards are named, but I guess, I can also look in the core node definitions, like you showed me.

**Comment by weathon** [comment]
Is there any documents for each datatype and functions in ComfyUI?

**Comment by schoenid** [comment]
> Is there any documents for each datatype and functions in ComfyUI?

The official documentation is https://docs.comfy.org/custom-nodes/overview and below.

Properties: https://docs.comfy.org/custom-nodes/backend/server_overview
Datatypes: https://docs.comfy.org/custom-nodes/backend/datatypes

JavaScript Extensions: https://docs.comfy.org/custom-nodes/js/javascript_overview
Hooks: https://docs.comfy.org/custom-nodes/js/javascript_hooks
Objects: https://docs.comfy.org/custom-nodes/js/javascript_objects_and_hijacking

Worth looking for:
https://github.com/chrisgoringe/Comfy-Custom-Node-How-To/wiki/aaa_index
https://github.com/comfyanonymous/ComfyUI/tree/master/comfy/comfy_types
https://github.com/cozy-comfyui
https://github.com/Comfy-Org/litegraph.js

...
also have a look at the standard nodes and other custom nodes.
